lib/worker/output_test.go

```diff
@@ -0,0 +1,77 @@
+package worker_test
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/trivelle/worker/lib/worker"
+)
+
+func TestOutputHandlerEmptyReader(t *testing.T) {
+	var stdoutReader bytes.Buffer
+
+	outputHandler, err := worker.NewOutputHandler(&stdoutReader)
+	assert.Nil(t, err)
+	outputChan, _ := outputHandler.Stream()
+
+	expectedOutputLines := []string{}
+
+	var outputLines []string
+	for out := range outputChan {
+		outputLines = append(outputLines, string(out.Content))
+	}
+
+	assert.ElementsMatch(t, outputLines, expectedOutputLines)
+}
+
+func TestOutputHandlerSuccess(t *testing.T) {
+	var stdoutReader bytes.Buffer
+	var stdErrReader bytes.Buffer
+
+	_, err := stdoutReader.WriteString("some output line 1\nsome output line 2\nsome output line 3")
+	assert.Nil(t, err, "error setting up reader")
+
+	_, err = stdErrReader.WriteString("some error")
+	assert.Nil(t, err, "error setting up reader")
+
+	outputHandler, err := worker.NewOutputHandler(&stdoutReader, &stdErrReader)
+	assert.Nil(t, err)
+	outputChan, _ := outputHandler.Stream()
+
+	expectedOutputLines := []string{"some output line 1\nsome output line 2\nsome output line 3", "some error"}
+
+	var outputLines []string
+	for out := range outputChan {
+		outputLines = append(outputLines, string(out.Content))
+	}
+
+	assert.ElementsMatch(t, outputLines, expectedOutputLines)
+}
+
+// errorReader is a fake reader that throws errors for testing
+type errorReader struct {
+}
+
+func (r *errorReader) Read(p []byte) (n int, err error) {
+	return 0, fmt.Errorf("errorReader returns errors")
+}
+
+func TestOutputHandlerError(t *testing.T) {
+	var stdoutReader bytes.Buffer
```

Comment:
There's something up in this test, as I'm getting intermittent test timeouts here.

Reply:
Reasonably sure you are occasionally deadlocking on sends on the unbuffered error channel in this test.

Reply:
I was puzzled about why this is happening. I had to re-run my test many times to replicate this issue. I think you are right, in the case were the stderr reader gets read first this test deadlocks forever because of the for range in the test.

Reply:
If i change the test to instead do this:
```
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		for range outputChan {
			// do nothing, just consume output
		}
		wg.Done()
	}()

	go func() {
		err = <-errChan
		wg.Done()
	}()
	wg.Wait()

```
Then it works. This is bad though... One channel shouldn't be affecting the other.

Reply:
It looks like a send on an unbuffered channel is blocking  inside `iterListeners()`, which holds the `OutptutHandler()` lock.

Reply:
Is your suggestion here: https://github.com/trivelle/worker/pull/2#discussion_r859489615

To instead do something like?
```
func (o *OutputHandler) getListeners() []*Listener {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.listeners[:len(o.listeners)]
}
```
and iterate on the value?

This seems to fix the issue in this test.

Reply:
I applied the change described above ^

Reply:
During the investigation of this issue I ended up finding an even more rare bug where a listener might end up with a duplicate message. This was occurring because the broadcasting and the buffering where happening in different steps which could result in a listener receiving both the `combinedBuffer` AND the new output chunk. Solved this by letting `handleBroadcast` do the buffering instead. I'd really like to revisit if I can simplify how the OutputHandler works to make it easier to understand and/or add more thorough tests but am not sure am gonna have the time.